Online DDL: managed online schema changed via gh-ost and pt-online-schema-change (#510)
vitessio/website: content/en/docs/user-guides/schema-changes/managed-online-schema-changes.md

---
title: Managed, Online Schema Changes
weight: 2
aliases: ['/docs/user-guides/managed-online-schema-changes/']
---

**Note:** this feature is **EXPERIMENTAL**. Also, the syntax for online-DDL is **subject to change**.

Vitess offers managed, online schema migrations, via [gh-ost](https://github.com/github/gh-ost) and [pt-online-schema-change](https://www.percona.com/doc/percona-toolkit/3.0/pt-online-schema-change.html). As a quick breakdown:

> - Vitess recognized a special `ALTER TABLE` syntax that indicates on online schema hcnage request.

Reviewer: ```suggestion
- Vitess recognizes a special `ALTER TABLE` syntax that indicates an online schema change request.
```

Author: fixed

> - Vitess responds to on online schema change request with a job ID

Reviewer: ```suggestion
- Vitess responds to an online schema change request with a job ID
```

Author: fixed

- Vitess resolves affected shards
- A shard's `primary` tablet schedules the migration to run when possible
- The tablets run migrations via `gh-ost` or `pt-online-schema-change`

> - Vitess provides the user the mechanism to show migration status, cancel or retry migrations, based on the job ID

Reviewer: ```suggestion
- Vitess provides the user a mechanism to view migration status, cancel or retry migrations, based on the job ID
```

Author: fixed

Welcome to the MySQL monitor.  Commands end with ; or \g.

mysql> ALTER WITH 'pt-osc' TABLE demo ADD COLUMN sample INT;
+--------------------------------------+
| uuid                                 |
+--------------------------------------+
| fa2fb689_f1d5_11ea_859e_f875a4d24e90 |
+--------------------------------------+
1 row in set (0.00 sec)
```

Just like in the previous example, `VTGate` identifies that this is an online schema change request, and persists it in global `topo`, returning a job ID for tracking. Migration does not start immediately.

## Migration flow and states

We highlight how Vitess manages migrations internally, and explain what states a migration goes through.

- Whether via `vtctlclient ApplySchema` or via `VTGate` as described above, a migration request entry is persisted in global `topo` (e.g. the global `etcd` cluster).
- `vtctld` periodically checks on new migration requests.
- `vtctld` resolves the relevant shards, and the `primary` tablet for each shard.
- `vtctld` pushes the request to all relevant `primary` tablets.
- If not all shards confirm receipt, `vtctld` periodically keeps retrying pushing the request to the shards until all approve.

> - Internally, tablet's persist the request on a designated table in the `_vt` schema. **Do not** manipulate that table directly as that would cause inconsistencies.

Reviewer: ```suggestion
- Internally, tablets persist the request in a designated table in the `_vt` schema. **Do not** manipulate that table directly as that can cause inconsistencies.
```

Author: fixed

- A shard's `primary` tablet owns running the migration. It is independent of other shards. It will schedule the migration to run when possible. A tablet will not run two migrations at the same time.
- A migration is first created in `queued` state.

> - If the tablet sees queued migration, and assuming there's no reason to hold up, it picks the oldest requested migration in `queued` state, and moves it to `ready` state.

Reviewer: ```suggestion
- If the tablet sees a queued migration, and assuming there's no reason to wait, it picks the oldest requested migration in `queued` state, and moves it to `ready` state.
```

Author: fixed

| test-0000000401 | c0-   | vt_commerce  | demo        | 2201058f_f266_11ea_bab4_0242c0a8b007 | gh-ost   | 2020-09-09 06:32:31 |                     | failed           |
| test-0000000301 | 80-c0 | vt_commerce  | demo        | 2201058f_f266_11ea_bab4_0242c0a8b007 | gh-ost   | 2020-09-09 06:32:31 |                     | failed           |
| test-0000000201 | 40-80 | vt_commerce  | demo        | 2201058f_f266_11ea_bab4_0242c0a8b007 | gh-ost   | 2020-09-09 06:32:31 |                     | failed           |
| test-0000000101 |   -40 | vt_commerce  | demo        | 2201058f_f266_11ea_bab4_0242c0a8b007 | gh-ost   | 2020-09-09 06:32:31 |                     | failed           |
+-----------------+-------+--------------+-------------+--------------------------------------+----------+---------------------+---------------------+------------------+
```

## Retrying a migration

The user may retry running a migration. If the migration is in `failed` or in `cancelled` state, Vitess will re-run the migration, with exact same arguments as previously intended. If the migration is in any other state, `retry` does nothing.

> It is not possible to retry a migration and with new variables. e.g. if the user initially runs `ALTER WITH 'gh-ost' '--max-load Threads_running=200' TABLE demo MODIFY id BIGINT` and the migration failed, it is not possible to retry with `'--max-load Threads_running=500'`.

Reviewer: ```suggestion
It is not possible to retry a migration with different parameters. e.g. if the user initially runs `ALTER WITH 'gh-ost' '--max-load Threads_running=200' TABLE demo MODIFY id BIGINT` and the migration failed, it is not possible to retry with `'--max-load Threads_running=500'`.
```

Reviewer: Or `options` instead of `parameters` or `variables`.

Author: fixed. `options` used.

- On other architectures:
  - Have `gh-ost` executable installed
  - Run `vttablet` with `-gh-ost-path=/full/path/to/gh-ost` flag

Vitess automatically creates a MySQL account for the migration, with a randomly generated password. The account is destroyed at the end of the migration.

Vitess takes care of setting up the necessary command line flags. It automatically creates a hooks directory and populates it with hooks that report `gh-ost`'s progress back to Vitess. You may supply additional flags for your migration as part of the `ALTER` statement. Examples:

- `ALTER WITH 'gh-ost' '--max-load Threads_running=200' TABLE demo MODIFY id BIGINT`
- `ALTER WITH 'gh-ost' '--critical-load Threads_running=500 --critical-load-hibernate-seconds=60' --default-retries=512 TABLE demo MODIFY id BIGINT`
- `ALTER WITH 'gh-ost' '--allow-nullable-unique-key --chunk-size 200' TABLE demo MODIFY id BIGINT`

Do not override the following flags: `alter, database, table, execute, max-lag, force-table-names, serve-socket-file, hooks-path, hooks-hint-token, panic-flag-file`.

> `gh-ost` throtlling is done via Vitess's own tablet throttler, based on replication lag. 

Reviewer: ```suggestion
`gh-ost` throttling is done via Vitess's own tablet throttler, based on replication lag. 
```

Author: fixed

Vitess automatically creates a MySQL account for the migration, with a randomly generated password. The account is destroyed at the end of the migration.

Vitess takes care of supplying the command line flags, the DSN, the username & password. It also sets up `PLUGINS` used to communicate migration progress back to the tablet. You may supply additional flags for your migration as part of the `ALTER` statement. Examples:

- `ALTER WITH 'pt-osc' '--null-to-not-null' TABLE demo MODIFY id BIGINT`
- `ALTER WITH 'pt-osc' '--max-load Threads_running=200' TABLE demo MODIFY id BIGINT`
- `ALTER WITH 'pt-osc' '--alter-foreign-keys-method auto --chunk-size 200' TABLE demo MODIFY id BIGINT`

Vitess tracks the state of the `pt-osc` migration. If it fails, Vitess makes sure to drop the migration triggers. Vitess keeps track of the migration even if the tablet itself restarts for any reason. Normally that would terminate the migration; vitess will cleanup the triggers if so, or will happily let the migration run to completion if not.

Do not override the following flags: `alter, pid, plugin, dry-run, execute, new-table-name, [no-]drop-new-table, [no-]drop-old-table`.

> `pt-osc` throtlling is done via Vitess's own tablet throttler, based on replication lag, and via a `pt-online-schema-change` plugin. 

Reviewer: ```suggestion
`pt-osc` throttling is done via Vitess's own tablet throttler, based on replication lag, and via a `pt-online-schema-change` plugin. 
```

Author: fixed
